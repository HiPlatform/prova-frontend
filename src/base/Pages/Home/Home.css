body {
  width: 100%;
  height: 100%;
  overflow-x: hidden;
  margin: 0;
  background-color: #303030;
  color: #fff; }

.homePage {
  width: 90%;
  max-width: 960px;
  margin: 0 auto;
  clear: both; }
