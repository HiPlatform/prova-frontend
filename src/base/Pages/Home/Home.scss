@import "../../../variables";

.homePage {
  width: 90%;
  max-width: 960px;
  margin: 0 auto;
  clear: both;
}