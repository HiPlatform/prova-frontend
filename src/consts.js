export const URLS = {
  SEARCH: 'https://api.spotify.com/v1/search',
};

export const CREDENTIALS = {
  headers: {
    Authorization: 'Bearer BQDGGAag6Hocue0iXDFWp0O9CdmjCSq6nI9R2-HE-_N-skmHG8V3btyKOZnilxWjn7cpkFMAbM70G8CH_-Y',
  },
};

export const HOMEPAGE_ACTIONS = {
  SEARCH_DATA: 'SEARCH_DATA',
  SEARCH_TYPE: 'SEARCH_TYPE',
  SEARCH_CHANGE: 'SEARCH_CHANGE',
  SEARCH_DETAILS: 'SEARCH_DETAILS',
  NAVIGATE: 'NAVIGATE',
};
