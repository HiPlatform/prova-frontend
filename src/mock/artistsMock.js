export default {
  artists: {
    items: [
      {
        genres: ['string folk'],
        href: 'https://api.spotify.com/v1/artists/5oW7xntz397IxAiDinC6Tk',
        images: [
          { url: 'https://i.scdn.co/image/4be2260d1e012c613f26d943b1788cc2a4c97e65' },
          { url: 'https://i.scdn.co/image/769dda96acd0d7fbf09b487fa34d283784fb3de8' },
          { url: 'https://i.scdn.co/image/9020f047835ed5ccac3e40ddef4ae15fbfa0cc1d' },
        ],
        name: 'John Mock',
        popularity: 26,
        type: 'artist',
      },
      {
        genres:['emo', 'emo punk'],
        href: 'https://api.spotify.com/v1/artists/5uDuKeTdnkTHdHIGrQWhOL',
        images: [
          { url: 'https://i.scdn.co/image/97bbdad7710788078378bd4d74538d6e88d120a9' },
          { url: 'https://i.scdn.co/image/789efd9cfba28a3b594084bf3be4d18061317201' },
          { url: 'https://i.scdn.co/image/d19e7fc70c44eec288fe518098fe9a853e3d97a0' },
          { url: 'https://i.scdn.co/image/ec87049d3415a82a575feaaf01d4aca6505b10f5' }
        ],
        name: 'Mock Orange',
        popularity: 23,
        type: 'artist',
      },
    ],
  },
};
