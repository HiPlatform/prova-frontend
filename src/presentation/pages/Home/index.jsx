import React from "react";
import { List } from "./components";

import Services from "../../../data/services";

const Home = () => {
	const data = new Services();
	return <List data={data.getData()} />;
};

export default Home;
