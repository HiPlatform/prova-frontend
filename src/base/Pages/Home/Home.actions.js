import { URLS, HOMEPAGE_ACTIONS } from '../../../consts';
import fetch from '../../../helpers/fetchData';

export const searchData =
  (type, text) =>
    dispatch =>
      fetch(
        `${URLS.SEARCH}/${type}/${text}`,
        HOMEPAGE_ACTIONS.SEARCH_DATA,
        'data',
        dispatch,
      );

export const changeType = searchType =>
  ({ type: HOMEPAGE_ACTIONS.SEARCH_TYPE, searchType });

export const changeInput = ({ target: { value } }) =>
  ({ type: HOMEPAGE_ACTIONS.SEARCH_CHANGE, value });
