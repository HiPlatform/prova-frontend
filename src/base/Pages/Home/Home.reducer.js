import { T, cond, propEq, pathOr } from 'ramda';

import { HOMEPAGE_ACTIONS, HOME_INITIALSTATE } from '../../../consts';

const getData = state => ({ data }) => {
  const items = pathOr([], [`${state.searchType}s`, 'items'], data);
  return { ...state, items };
};

const reducer = state => cond([
  [propEq('type', HOMEPAGE_ACTIONS.SEARCH_DATA), getData(state)],
  [propEq('type', HOMEPAGE_ACTIONS.SEARCH_TYPE), ({ searchType }) => ({ ...state, searchType })],
  [propEq('type', HOMEPAGE_ACTIONS.SEARCH_CHANGE), ({ value }) => ({ ...state, value })],
  [T, () => state],
]);

export default (state = HOME_INITIALSTATE, action) => reducer(state)(action);
