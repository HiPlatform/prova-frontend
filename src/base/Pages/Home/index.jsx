import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import { changeType, changeInput, searchData } from './Home.actions';
import HomeForm from '../../Templates/HomeForm';
import './Home.css';

export const Home = ({
  value,
  searchType,
  fnType,
  fnInput,
  fnSearch,
}) => (
  <div className="homePage">
    <HomeForm
      value={value}
      searchType={searchType}
      fnType={fnType}
      fnInput={fnInput}
      fnSearch={fnSearch}
    />
  </div>
);

Home.defaultProps = {
  value: '',
  searchType: '',
  fnType: () => {},
  fnInput: () => {},
  fnSearch: () => {},
};

Home.propTypes = {
  value: PropTypes.string,
  searchType: PropTypes.string,
  fnType: PropTypes.func,
  fnInput: PropTypes.func,
  fnSearch: PropTypes.func,
};

const mapState2ToProps = state => ({
  value: state.home.value,
  searchType: state.home.searchType,
});

const mapDispatch2Props = dispatch =>
  bindActionCreators({
    fnType: changeType,
    fnInput: changeInput,
    fnSearch: searchData,
  }, dispatch);

export default connect(mapState2ToProps, mapDispatch2Props)(Home);
