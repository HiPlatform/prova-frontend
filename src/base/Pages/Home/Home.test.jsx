import React from 'react';
import { shallow } from 'enzyme';
import { shallowWithStore } from 'enzyme-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { partial } from 'ramda';

import Connected, { Home } from './';
import { searchData, changeType, changeInput } from './Home.actions';
import HomeReducer from './Home.reducer';
import { HOME_INITIALSTATE, HOMEPAGE_ACTIONS } from '../../../consts';
import artistsMock from '../../../mock/artistsMock';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('HomePage >', () => {
  const wrapper = shallow(<Home />);
  const store = mockStore({ home: HOME_INITIALSTATE });
  const connectedWrapper = shallowWithStore(<Connected />, store);
  const event = { target: { value: 'some input' } };
  const { searchType } = connectedWrapper.props();
  const typeAction = changeType(searchType);
  const inputAction = changeInput(event);
  const searchAction = searchData(searchType, 'mock');
  const randomAction = { type: 'random' };

  it('should render without crashing', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('[Connected] HomePage >', () => {
    beforeEach(() => {
      store.clearActions();
    });

    it('should the prop searchType value be equal "artist"', () => {
      expect(connectedWrapper.props().searchType).toBe(HOME_INITIALSTATE.searchType);
    });

    it('dispatch action when user clicks in one of the buttons of the GroupButtons', () => {
      const value = 'tracks';
      const expectedResult = changeType(value);

      connectedWrapper.props().fnType(value);
      expect(store.getActions()).toEqual([expectedResult]);
    });

    it('dispatch action when user changes the input', () => {
      connectedWrapper.props().fnInput(event);
      expect(store.getActions()).toEqual([inputAction]);
    });
  });

  describe('[Actions] HomePage >', () => {
    describe('searchData >', () => {
      beforeEach(() => {
        store.clearActions();
        fetch.resetMocks();
        fetch.mockResponse(JSON.stringify(artistsMock));
      });

      it('should call the api and get the "searchType" data', () => {
        const expectedResult = [{ type: HOMEPAGE_ACTIONS.SEARCH_DATA, data: artistsMock }];

        return store.dispatch(searchAction).then(() => {
          expect(store.getActions()).toEqual(expectedResult);
        });
      });
    });

    describe('changeType >', () => {
      it('should return the selected searchType', () => {
        const expectedResult = { type: HOMEPAGE_ACTIONS.SEARCH_TYPE, searchType };

        expect(typeAction).toEqual(expectedResult);
      });
    });

    describe('changeInput >', () => {
      it('should return the typed value of the input', () => {
        const expectedResult = { type: HOMEPAGE_ACTIONS.SEARCH_CHANGE, value: 'some input' };

        expect(inputAction).toEqual(expectedResult);
      });
    });
  });

  describe('[Reducer] HomePage >', () => {
    const reducer = partial(HomeReducer, [HOME_INITIALSTATE]);

    it('has default called', () => {
      expect(reducer(randomAction)).toEqual(HOME_INITIALSTATE);
    });

    it('sets { searchType: "track" } when called with type action and "track" as value', () => {
      const modifiedTypeAction = { ...typeAction, searchType: 'track' };
      const expectedResult = { ...HOME_INITIALSTATE, searchType: 'track' };

      expect(reducer(modifiedTypeAction)).toEqual(expectedResult);
    });

    it('sets { value: "mock" } when called with input action and "mock" as value', () => {
      const modifiedInputAction = { ...inputAction, value: 'mock' };
      const expectedResult = { ...HOME_INITIALSTATE, value: 'mock' };

      expect(reducer(modifiedInputAction)).toEqual(expectedResult);
    });

    it('sets items based on search attributes when called with search action', () => {
      const modifiedSearchAction = { type: HOMEPAGE_ACTIONS.SEARCH_DATA, data: artistsMock };
      const expectedResult = { ...HOME_INITIALSTATE, items: artistsMock.artists.items };

      expect(reducer(modifiedSearchAction)).toEqual(expectedResult);
    });
  });
});
