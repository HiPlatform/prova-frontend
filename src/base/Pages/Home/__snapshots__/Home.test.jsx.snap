// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage  should render without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    fnInput={[Function]}
    fnSearch={[Function]}
    fnType={[Function]}
    searchType=""
    value=""
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <HomeForm
        fnInput={[Function]}
        fnSearch={[Function]}
        fnType={[Function]}
        searchType=""
        value=""
/>,
      "className": "homePage",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "fnInput": [Function],
        "fnSearch": [Function],
        "fnType": [Function],
        "searchType": "",
        "value": "",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <HomeForm
          fnInput={[Function]}
          fnSearch={[Function]}
          fnType={[Function]}
          searchType=""
          value=""
/>,
        "className": "homePage",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "fnInput": [Function],
          "fnSearch": [Function],
          "fnType": [Function],
          "searchType": "",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;

exports[`HomePage > should render without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    fnInput={[Function]}
    fnSearch={[Function]}
    fnType={[Function]}
    searchType=""
    value=""
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <HomeForm
        fnInput={[Function]}
        fnSearch={[Function]}
        fnType={[Function]}
        searchType=""
        value=""
/>,
      "className": "homePage",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "fnInput": [Function],
        "fnSearch": [Function],
        "fnType": [Function],
        "searchType": "",
        "value": "",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <HomeForm
          fnInput={[Function]}
          fnSearch={[Function]}
          fnType={[Function]}
          searchType=""
          value=""
/>,
        "className": "homePage",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "fnInput": [Function],
          "fnSearch": [Function],
          "fnType": [Function],
          "searchType": "",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;

exports[`HomePage should render without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    fnInput={[Function]}
    fnSearch={[Function]}
    fnType={[Function]}
    searchType=""
    value=""
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <HomeForm
        fnInput={[Function]}
        fnSearch={[Function]}
        fnType={[Function]}
        searchType=""
        value=""
/>,
      "className": "homePage",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "fnInput": [Function],
        "fnSearch": [Function],
        "fnType": [Function],
        "searchType": "",
        "value": "",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <HomeForm
          fnInput={[Function]}
          fnSearch={[Function]}
          fnType={[Function]}
          searchType=""
          value=""
/>,
        "className": "homePage",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "fnInput": [Function],
          "fnSearch": [Function],
          "fnType": [Function],
          "searchType": "",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
